### ISSUE TITLE
Potential over validation of OIDC URI

### ISSUE THREAD

**Issue by Firstyear** [IdentityModel/IdentityModel#553]
A number of Kanidm's community users have been attempting to use JellyFin with Kanidm. JellyFin's OIDC support is provided through this IdentityModel library.

When used users would encounter the error `Endpoint is on a different host than authority`. However, Kanidm does not allow multiple hosts/origins on an oauth2 provider. 

The discover output as example is:

```
https://auth.domain.tld/oauth2/openid/jellyfin-aer/.well-known/openid-configuration
```

```
{
  "issuer": "https://auth.domain.tld/oauth2/openid/jellyfin-aer",
  "authorization_endpoint": "https://auth.domain.tld/ui/oauth2",
  "token_endpoint": "https://auth.domain.tld/oauth2/token",
  "userinfo_endpoint": "https://auth.domain.tld/oauth2/openid/jellyfin-aer/userinfo",
  "jwks_uri": "https://auth.domain.tld/oauth2/openid/jellyfin-aer/public_key.jwk",
  "scopes_supported": [
    "openid",
    "profile"
  ],
  "response_types_supported": [
    "code"
  ],
  "response_modes_supported": [
    "query"
  ],
  "grant_types_supported": [
    "authorization_code"
  ],
  "subject_types_supported": [
    "public"
  ],
  "id_token_signing_alg_values_supported": [
    "ES256"
  ],
  "token_endpoint_auth_methods_supported": [
    "client_secret_basic",
    "client_secret_post"
  ],
  "display_values_supported": [
    "page"
  ],
  "claim_types_supported": [
    "normal"
  ],
  "service_documentation": "https://kanidm.github.io/kanidm/master/integrations/oauth2.html",
  "claims_parameter_supported": false,
  "request_parameter_supported": true,
  "request_uri_parameter_supported": false,
  "require_request_uri_registration": false
}
```

As we can see, all of the URIs are "https://auth.domain.tld". We believe that the issue is over validation that all URIs are relative / beneath the issuer URI.

Per https://openid.net/specs/openid-connect-discovery-1_0.html does not require relative locations of urls. In fact it seems the only requirement is that iss is the prefix to .well-known (section 4.3), which we do correctly.

We do however agree the this validation has value ( https://github.com/IdentityModel/IdentityModel/issues/456 ). Verification that all of these URIs have the same scheme and origin is a valuable defence. However, enforcing that all of these URIs are relative to issuer creates issues with IDPs as above without a measurable increase in security as users now disable the verification setting instead. We do not want Jellyfin or Kanidm to have to recommend disabling this verification. 

We would like to suggest that 

* IdentityModels `ValidateEndpoints` is relaxed to only enforce scheme and origin of the URI match, which matches the intent of https://github.com/IdentityModel/IdentityModel/issues/456  in our view
* That the error return it's expected values, not just the endpoint value (e.g. `Endpoint is on a different host than authority: {endpoint} expected {host}` )

CC: @9p4 

Related to: https://github.com/9p4/jellyfin-plugin-sso/discussions/156

**Comment by leastprivilege** [comment]
It's been a while, but you know that you can plugin your own validators?

**Comment by Firstyear** [comment]
That may be so, but the defaults matter and if they are "over-verifying" then people will disable them. So shouldn't it do the right things by default?

**Comment by leastprivilege** [comment]
That would be up for discussion - it is always hard to change defaults in a library that has hundreds of millions of downloads. Feel free to write up a proposal in form of a PR.

**Comment by Firstyear** [comment]
I'm not a C# developer, I'm the person who writes the IDP that's triggering this behaviour. I raised this to hopefully help and improve the defaults you offer, but I won't be submitting any PR's. That would be up to @9p4 or yourself.

**Comment by brockallen** [comment]
> We would like to suggest that
> That the error return it's expected values, not just the endpoint value (e.g. Endpoint is on a different host than authority: {endpoint} expected {host} )

@josephdecock I could see adding this requested change in our current round of work. 

I'm unsure if we should change the behavior of the rest. My sense is that no one would know enough to ever enable the strong validation check.

**Comment by josephdecock** [comment]
The spec doesn't require validation of paths, but it does specifically allow paths to be part of the issuer to support multi tenancy. If you make the assumption that such multi tenant issuers would organize their endpoints underneath the issuer, I think there is value in validating the paths. And while that is an assumption/not explicitly part of the spec, I would be reluctant to change the default behavior.

What if we kept the current behavior as default, but add a flag to disable validation of paths?

**Comment by 9p4** [comment]
There is already a flag to disable the validation of paths, but this is not ideal since this issue is about the same domain (with different paths).

**Comment by Firstyear** [comment]
This would resolve our issues for our users, while retaining the origin validation that is a valuable security check.

**Comment by Firstyear** [comment]
> There is already a flag to disable the validation of paths.

It also disables origin checking.

**Comment by josephdecock** [comment]
Looking at this a little closer, there actually already is the `DiscoveryPolicy.AdditionalEndpointBaseAddresses` that will accomplish this. I verified the usage by putting the sample discovery document above into the file discovery_endpoints_not_relative.json, and then running this test (which passes without changes):

```cs
        [Fact]
        public async Task AdditionalEndpointBaseAddresses_should_be_respected()
        {
            var discoFileName = FileName.Create("discovery_endpoints_not_relative.json");
            var document = File.ReadAllText(discoFileName);
            var handler = new NetworkHandler(request => document, HttpStatusCode.OK);
            var client = new HttpClient(handler)
            {
                BaseAddress = new Uri("https://auth.domain.tld/oauth2/openid/jellyfin-aer")
            };

            var discoRequest = new DiscoveryDocumentRequest();
            discoRequest.Policy.AdditionalEndpointBaseAddresses.Add("https://auth.domain.tld");

            var disco = await client.GetDiscoveryDocumentAsync(discoRequest);

            disco.IsError.Should().BeFalse();
            disco.AuthorizeEndpoint.Should().Be("https://auth.domain.tld/ui/oauth2");
            disco.TokenEndpoint.Should().Be("https://auth.domain.tld/oauth2/token");
            disco.JwksUri.Should().Be("https://auth.domain.tld/oauth2/openid/jellyfin-aer/public_key.jwk");
        }
```

So, I think we don't need to add any new options here. We can just update the error message to help users understand the situation better, and maybe even mention that AdditionalEndpointBaseAddresses exists.

**Comment by josephdecock** [comment]
The new error message is now released in preview 3 of our upcoming 7.0.0 release!
